### PR TITLE
Update product stock status in Facebook catalog when stock changes in the store

### PR DESCRIPTION
**Pull request by wvega**
# Summary

This PR schedules products for sync or delete when WooCommerce increases or decreases the product stock.

## Details

WooCommerce routes all product stock changes through the `wc_update_product_stock()` function. This PR adds callbacks for the `woocommerce_variation_set_stock` and `woocommerce_product_set_stock` hooks to schedule products for sync as soon as their stock is increased or decreased.

## UI Changes

None

### Setup

1. Connect your store to Facebook
1. Create a Simple Product with Facebook sync enabled
    1. Enable **Manage stock?** and set the stock quantity to 1
    1. Define a price for the product
1. Create a Variable product
    1. Enable **Manage stock?** at the Variable product level and set the stock quantity to 1
    1. Make sure the product has at least two variations with a price and Facebook sync enabled
    1. Enable **Manage stock?** at the variation level for one of the variations and set the stock quantity to 1. Lets call this variation the **Variation with Stock**.

#### Simple Product

1. Go to the Product Catalog on Facebook
     - [x] The Simple product's availability is set to `In stock`
1. Add the simple product to the cart and place an order
1. Ensure the Order status is one of On Hold, Processing, or Completed
1. Go to the Product Catalog on Facebook
     - [x] The Simple products availability is set to `Out of stock`
1. Cancel the order
1. Go to the Product Catalog on Facebook
     - [x] The Simple products availability is set to `In stock`

#### Variable Product

1. Go to the Product Catalog on Facebook
     - [x] The availability for all product variants is set to `In stock` (click the product card, then the **Variants** tab)
1. Add the **Variation with Stock** product to the cart and place an order
1. Ensure the Order status is one of On Hold, Processing, or Completed
1. Go to the Product Catalog on Facebook
     - [x] The **Variation with Stock** availability is set to `Out of stock`
     - [x] The availability for the other variants is set to `In stock`
1. Add one of the other variations to the cart and place an order
1. Ensure the Order status is one of On Hold, Processing, or Completed
1. Go to the Product Catalog on Facebook
     - [x] The **Variation with Stock** availability is set to `Out of stock`
     - [x] The availability for the others variants is set to `Out of stock`
1. Cancel the order for the **Variation with Stock**
1. Go to the Product Catalog on Facebook
     - [x] The **Variation with Stock** availability is set to `In stock`
     - [x] The availability for the others variants is set to `Out of stock`

## Before merge

- [x] I have confirmed these changes in each supported minor WooCommerce version

### CODE REVIEW

File: includes/Products/Stock.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		$this->add_hooks();
+	}
+
+
+	/**
+	 * Adds needed hooks to keep product stock in sync.
+	 *
+	 * @since 2.0.4-dev.1
+	 */
+	public function add_hooks() {
```
Comment (unfulvio) [inline includes/Products/Stock.php:41]:
```suggestion
	private function add_hooks() {
```

is there a reason for this to be public?

**Review by unfulvio — APPROVED** [review]
hey @wvega this seemed to work initially; code checks out, integration tests past. I noticed though that FB tends to hide products out of stock from my catalog, after they've been marked out of stock. I'm not sure why is that. Is there a setting somewhere?

**Comment by wvega** [comment]
@unfulvio thank you for taking a look.

>I noticed though that FB tends to hide products out of stock from my catalog, after they've been marked out of stock. I'm not sure why is that. Is there a setting somewhere?

You mean they don't show up in the Page's shop? I'm not exactly sure if there is a setting to change that. As far as I can tell all we can do is set the availability.

If you have WooCommerce > Settings > Products > Inventory > Hide out of stock items from the catalog enabled, the plugin will issue a request to remove the product item from the catalog instead of updating its info. Do you have that option enabled?

File: includes/Products/Stock.php (hunk 1 above)
Comment (wvega) [reply includes/Products/Stock.php:41]: Not really. It was copy pasta.

**Comment by unfulvio** [comment]
> If you have WooCommerce > Settings > Products > Inventory > Hide out of stock items from the catalog enabled, the plugin will issue a request to remove the product item from the catalog instead of updating its info. Do you have that option enabled?

oh, you're right! I had this setting enabled, now everything looks ok!

**Comment by unfulvio** [comment]
@wvega after confirming, I've fixed a merge conflict - seems 2.0.4 is out but `release-2.0.5` needs to pick up some changes first?

**Comment by wvega** [comment]
@unfulvio you're right. We merged another #1624 into `release-2.0.4`  instead of `release-2.0.5`.

I fixed a couple of issues with line endings from #1624 and then [merged `release-2.0.4` into `release-2.0.5`](https://github.com/facebookincubator/facebook-for-woocommerce/commit/3dd2335c2cc11ef72331c09aafc73b93479b27d7).

The branch should be up to date now. Thanks!

**Pull request merged**